### ISSUE TITLE
Implement Typed Function References Proposal for WebAssembly

### ISSUE THREAD

**Issue by fniephaus** [oracle/graal#9760]
### TL;DR

We plan to implement [the WebAssembly Typed Function References proposal](https://github.com/WebAssembly/function-references) in GraalWasm.

### Details

This proposal adds function references that are typed and can be called directly.
Unlike the existing `funcref` type and `call_indirect` instruction, typed function references do not need to be stored in a table to be called. This simplifies interoperability with the host, making it easier to pass callbacks from Java or other Graal Languages to WebAssembly code. Since these function references will be typed, this also means that indirect function calls within a WebAssembly module can now be executed with fewer runtime checks.

More details at https://github.com/WebAssembly/gc/blob/main/proposals/function-references/Overview.md
